File: ex1_kmeans.py
import math

# global variables:
c_arr = [[]] # centroinds array [k*d]
p_arr = [[]] # points array [N*d]
K, N, d = 0, 0, 0


epsilon = 0.001
iterations = 200


def kmeans():
    '''Implement k-means clustering algorithm'''

    global p_arr, c_arr, K, N, d
    
    # to do

    # example:
    p_arr = [[1,2],[3,4],[5,6]]
    c_arr = [[0,0],[1,1]]
    d=2
    N=3
    K=2
    
    for i in range(iterations):
        a_i_dict = assign() # a_i_dict := {cluster index : [point indices]}
        delta_c = update(a_i_dict) # delta_c := maximum distance between the prev. and cur. centroids 
        if convergence(delta_c):
            break


def assign():
    '''Assign every point to the closest cluster'''

    a_i_dict = {i:[] for i in range(K)}
    for i in range(N):
        min_index = argmin_distance(c_arr,p_arr[i])
        a_i_dict[min_index].append(i) # optional: we can use an array instead.
    return a_i_dict


def update(a_i_dict):
    '''Update the centroids' coordinates to be the center of mass of their associated points'''

    delta_c = 0
    for i in range(K):
        length = len(a_i_dict[i])
        if length!=0:
            new_c = [sum([p_arr[a_i_dict[i][j]][m] for j in range(length)])/length for m in range(d)]
            delta_c = max(distance_pp(new_c,c_arr[i]),delta_c)
            c_arr[i] = new_c
    return delta_c


def convergence(delta_c):
    '''Check if the condition for convergence is met'''

    if delta_c < epsilon:
        return 1
    return 0


def distance_pp(p1,p2):
    '''Calculate the euclidean distance between two points in R^d'''

    return math.sqrt(sum([(p2[i]-p1[i])**2 for i in range(d)]))


def argmin_distance(c_arr,p):
    '''Input: c_arr-list of centroids [k*d], p-point in R^d. 
    Maintaining the minimum index from p to the centroids.'''

    min_index = 0
    d_min = distance_pp(c_arr[0],p)
    for i in range(K):
        d_cur = distance_pp(c_arr[i],p)
        if d_cur<d_min:
            min_index = i
            d_min = d_cur
    return min_index



     

File: ex1_kmeans.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>

/*---Global Variables---*/
double epsilon = 0.001;
int iterations = 200;
double delta_c = 0;
int K;
long N;
int d;
double **p_arr; /*points array [N*d]*/
double **c_arr; /*centroinds array [k*d]*/ 
double **c_sum_arr; /*centroinds sum array [k*d]*/
int *c_size_arr; /*size of each cluster array [k*1]*/

int main();
void kmeans();
void assign();
void update();
char convergence();
double distance_pp(double*, double*);
int argmin_distance(double**, double*);
void vec_sum(double*, double*);
double* vec_div(double*, int);

int main()
{
    /*to do*/
    kmeans();
    return 0;
}

void kmeans()
{
    c_size_arr = (int*)calloc(K, sizeof(int));
    c_sum_arr = (double**)calloc(K, sizeof(double*));

    /*...*/

    /*example:*/
    double p_arr[3][2] = {{1,0}, {3,0}, {5,0}};
    double c_arr[2][2] = {{0,0},{1,1}};
    d=2;
    N=3;
    K=2;

    int i;
    for(i=0; i<iterations; i++)
    {
        assign(); /*a_i_dict := {cluster index : [point indices]}*/
        update(); /*delta_c := maximum distance between the prev. and cur. centroids*/ 
        if (convergence())
        {
           break; 
        }
            
    }

    free(c_size_arr);
    free(c_sum_arr);
}


void assign(int n, points, centrieds)
{
    /*Assign every point to the closest cluster*/

    int i;
    int min_index;
    
    memset(c_sum_arr, 0, sizeof(c_sum_arr));
    memset(c_size_arr, 0, sizeof(c_size_arr));
   
    for(i = 0; i < N; i++)
    {
        min_index = argmin_distance(c_arr,p_arr[i]);
        vec_sum(c_sum_arr[min_index], p_arr[i]);
        c_size_arr[min_index] += 1;
    }
}


void update()
{
    /*Update the centroids coordinates to be 
    the center of mass of their associated points*/
    
    delta_c = 0;
    int length;
    int i;
    double* new_c = calloc(d, sizeof(int));

    for(i = 0; i < K; i++)
    {
        length = c_size_arr[i];
        if (length!=0)
        {
            new_c = vec_div(c_sum_arr[i],length);
            delta_c = fmax(distance_pp(new_c,c_arr[i]),delta_c);
            c_arr[i] = new_c;   
        }
    }
    free(new_c);
}


char convergence()
{
    /*Check if the condition for convergence is met*/
    if (delta_c < epsilon)
    {
        return 1;
    }  
    return 0;
}



double distance_pp(double *p1, double *p2)
{
    /*Calculate the euclidean distance between two points in R^d*/
    int i;
    double dist= 0.0;
    for(i = 0; i < d; i++)
    {
        dist += pow((p1[i] - p2[i]) , 2);
    }
    dist = sqrt(dist);
    return dist;
}


int argmin_distance(double **a, double *p)
{
    /*Input: c_arr-list of centroids [k*d], p-point in R^d. 
    Maintaining the minimum index from p to the centroids*/
    int i;
    int min_index = 0;
    double d_min = distance_pp(a[0],p);
    double d_cur;
    for(i = 0; i < K; i++)
    {
        d_cur = distance_pp(a[i],p);
        if (d_cur<d_min)
        {
            min_index = i;
            d_min = d_cur;
        }
    }
    return min_index;
}


void vec_sum(double *v1, double *v2)
{
    /*Sum v2 into v1, each component seperately*/

    int i;
    for(i = 0; i < d; i++)
    {
        v1[i] += v2[i];
    }
}


double *vec_div(double *v1, int l)
{
    /*Divide v1 by l, each component seperately*/

    int i;
    for (i = 0; i < d; i++)
    {
        v1[i] = (double)(v1[i] / l);
    }
    return v1;
}